### PR TITLE
CM APP 60 - Lock user if the user enters invalid password while changing password

### PR DESCRIPTION
**Pull request by gayatrijagtap**

### CODE REVIEW

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java [hunk 1]
```diff
@@ -25,12 +25,16 @@
...
-                                ? Mono.empty() : removeLockedUser(cmId))
-                .then(lockedUsersRepository.upsert(cmId));
+                .flatMap(lockedUser -> {
+                    var remainingTries = lockedServiceProperties.getMaximumInvalidAttempts() - lockedUser.getInvalidAttempts();
+                    return isBeforeMinutes(lockedUser.getDateCreated(), lockedServiceProperties.getCoolOfPeriod())
+                            ? Mono.just(remainingTries - 1)
+                            : removeLockedUser(cmId).then(Mono.just(lockedServiceProperties.getMaximumInvalidAttempts() - 1));
+                })
+                .flatMap(remainingTries -> lockedUsersRepository.upsert(cmId).thenReturn(remainingTries))
+                .switchIfEmpty(Mono.defer(() -> lockedUsersRepository.upsert(cmId).then(Mono.just(lockedServiceProperties.getMaximumInvalidAttempts() - 1))));
```
Comment (dheeraj-p) [inline src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:37]: We can use `thenReturn` here. `then(Mono.just(lockedServiceProperties.getMaximumInvalidAttempts() - 1))`

Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:37]: yeah off course. will change it

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  1
- src/main/java/in/projecteka/consentmanager/user/LockedUserService.java  1
- src/main/java/in/projecteka/consentmanager/clients/ClientError.java  1
- src/main/java/in/projecteka/consentmanager/user/UserService.java  3
- src/test/java/in/projecteka/consentmanager/user/LockedUserServiceTest.java  1
- src/main/java/in/projecteka/consentmanager/user/PatientsController.java  5
- src/test/java/in/projecteka/consentmanager/user/PatientControllerTest.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  13
- src/test/java/in/projecteka/consentmanager/user/PatientControllerTest.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5601226&bid=18349812)

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java [hunk 2]
```diff
@@ -25,12 +25,16 @@
...
-                                ? Mono.empty() : removeLockedUser(cmId))
-                .then(lockedUsersRepository.upsert(cmId));
+                .flatMap(lockedUser -> {
+                    var remainingTries = lockedServiceProperties.getMaximumInvalidAttempts() - lockedUser.getInvalidAttempts();
+                    return isBeforeMinutes(lockedUser.getDateCreated(), lockedServiceProperties.getCoolOfPeriod())
+                            ? Mono.just(remainingTries - 1)
+                            : removeLockedUser(cmId).thenReturn(lockedServiceProperties.getMaximumInvalidAttempts() - 1);
+                })
+                .flatMap(remainingTries -> lockedUsersRepository.upsert(cmId).thenReturn(remainingTries))
+                .switchIfEmpty(Mono.defer(() -> lockedUsersRepository.upsert(cmId).thenReturn(lockedServiceProperties.getMaximumInvalidAttempts() - 1)));
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:37]: lockedUsersRepository.upsert(cmId) this is common right? why can't you do it in the initial step itself.

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java [hunk 3]
```diff
@@ -25,12 +25,16 @@
...
         return lockedUsersRepository.getLockedUserFor(cmId)
-                .flatMap(lockedUser ->
-                        isBeforeMinutes(lockedUser.getDateCreated(), lockedServiceProperties.getCoolOfPeriod())
-                                ? Mono.empty() : removeLockedUser(cmId))
-                .then(lockedUsersRepository.upsert(cmId));
+                .flatMap(lockedUser -> {
+                    var remainingTries = lockedServiceProperties.getMaximumInvalidAttempts() - lockedUser.getInvalidAttempts();
+                    return isBeforeMinutes(lockedUser.getDateCreated(), lockedServiceProperties.getCoolOfPeriod())
+                            ? Mono.just(remainingTries - 1)
+                            : removeLockedUser(cmId).thenReturn(lockedServiceProperties.getMaximumInvalidAttempts() - 1);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:34]: whenever you have ternary operator, that's a sign for possible filter operator.

File: src/main/java/in/projecteka/consentmanager/user/PatientsController.java [hunk 4]
```diff
@@ -195,7 +196,10 @@ private OtpMediumType getOtpMedium(UserSignUpEnquiry userSignUpEnquiry) {
                 ? ReactiveSecurityContextHolder.getContext()
                 .map(securityContext -> (Caller) securityContext.getAuthentication().getPrincipal())
                 .map(Caller::getUsername)
-                .flatMap(userName -> userService.updatePassword(request, userName))
+                .flatMap(userName ->
+                                lockedUserService.validateLogin(userName)
+                                .then(Mono.defer(() -> userService.updatePassword(request, userName)))
+                        )
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/user/PatientsController.java:202]: why can't the .then be outside of flatmap?

File: src/main/java/in/projecteka/consentmanager/clients/ClientError.java [hunk 5]
```diff
@@ -323,4 +323,8 @@ public static ClientError noPatientFound() {
     public static ClientError invalidResponseFromHIP() {
         return new ClientError(UNPROCESSABLE_ENTITY, new ErrorRepresentation(new Error(UNPROCESSABLE_RESPONSE_FROM_GATEWAY, "Could not process response from HIP")));
     }
+
+    public static ClientError invalidOldPassword(int remainingAttempts) {
+        return new ClientError(UNAUTHORIZED, new ErrorRepresentation(new Error(USERNAME_OR_PASSWORD_INCORRECT,"You have " + remainingAttempts + " tries available after that your account will be locked")));
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/clients/ClientError.java:328]:
```suggestion
        return new ClientError(UNAUTHORIZED, new ErrorRepresentation(new Error(USERNAME_OR_PASSWORD_INCORRECT,"You have " + remainingAttempts + " tries available after that your account will be locked")));
```
```suggestion
        return new ClientError(UNAUTHORIZED, new ErrorRepresentation(new Error(USERNAME_OR_PASSWORD_INCORRECT, "You have " + remainingAttempts + " tries available after that your account will be locked")));
```

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java (hunk 2 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:37]: if we do so, the above calculations will get wrong as the upsert will update the user.

File: src/main/java/in/projecteka/consentmanager/user/PatientsController.java (hunk 4 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/PatientsController.java:202]: but then, it won't be able to access userName in then

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java (hunk 3 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:34]: not sure how can I do that in this case. If you have any suggestion, please tell me

File: src/main/java/in/projecteka/consentmanager/user/PatientsController.java (hunk 4 above)
Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/user/PatientsController.java:202]: then return username from validateLogin if username is valid. then do flatMap

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java (hunk 3 above)
Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:34]: in consent manager.cs we are doing it in many places. Refer those please. If you can't, ignore this.

Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:34]: .cs in c#? which service?

File: src/main/java/in/projecteka/consentmanager/user/PatientsController.java (hunk 4 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/PatientsController.java:202]: yeah. will do that

File: src/main/java/in/projecteka/consentmanager/user/LockedUserService.java (hunk 3 above)
Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/user/LockedUserService.java:34]: sorry ConsentManager.java

**Pull request merged**